docs/language/ql-training-rst/cpp/bad-overflow-guard.rst

```diff
@@ -0,0 +1,228 @@
+  - An overflow is when an arithmetic operation, such as an addition, results in a number which is too large to be stored in the type.
+  - When an operation overflows, the value “wraps” around.
+  - A typical way to check for overflow of an addition, therefore, is whether the result is less than one of the arguments - that is the result has **wrapped**.
+
+Integer promotion
+=================
+
+From `https://en.cppreference.com/w/c/language/conversion <https://en.cppreference.com/w/c/language/conversion>`__:
+
+*Integer promotion is the implicit conversion of a value of any integer type with rank less or equal to rank of* ``int`` *... to the value of type* ``int`` *or* ``unsigned int``.
+
+The arguments of the following arithmetic operators undergo implicit conversions:
+
+- binary arithmetic (* / % + - )
+- relational operators (< > <= >= == !=)
+- binary bitwise operators (& ^ \|)
+- the conditional operator (?:)
+
+.. note::
+
+  - Most of the time integer conversion works fine. However, the rules governing addition in C/C++ are complex, and it easy to get bitten.
+  - CPUs generally prefer to perform arithmetic operations on 32 bit or larger integers, as the architectures are optimised to perform those efficiently.
+  - The compiler therefore performs “integer promotion” for arguments to arithmetic operations that are smaller than 32 bit.
+
+Checking for overflow in C revisited
+====================================
+
+Which branch gets executed in this example? What is the value of ``result``?
+
+.. code-block:: cpp
+
+  uint16_t v = 65535;
+  uint16_t b = 1;
+  uint16_t result;
+  if (v + b < v) {
+      handle_error("overflow");
+  } else {
+      result = v + b;
+  }
+
+.. note::
+
+  In this example the second branch is executed, even though there is a 16-bit overflow, and ``result`` is set to zero.
+
+Checking for overflow in C revisited
+====================================
+
+Here is the example again, with the conversions made explicit:
+
+.. code-block:: cpp
+
+  uint16_t v = 65535;
+  uint16_t b = 1;
+  uint16_t result;
+  if ((int)v + (int)b < (int)v) {
+      handle_error("overflow");
+  } else {
+      result = (uint16_t)((int)v + (int)b);
+  }
+
+.. note::
+
+  In this example the second branch is executed, even though there is a 16-bit overflow, and result is set to zero.
+
+  Explanation:
+  - The two integer arguments to the addition, v and b, are promoted to 32 bit integers.
+  - The comparison (<) is also an arithmetic operation, therefore it will also be completed on 32 bit integers.
+  - This means that v + b < v will never be true, because v and b can hold at most 216.
+  - Therefore, the second branch is executed, but the result of the addition is stored into the result variable. Overflow will still occur as result is a 16 bit integer.
+
+This happens even though the overflow check passed!
+
+.. rst-class:: background2
+
+Developing a QL query
+=====================
+
+Finding bad overflow guards
```

Comment:
The original slide has notes.

Reply:
Yep. I left these notes on purpose. I'll discuss this with Luke.